src/binary_format.cc

```diff
@@ -0,0 +1,254 @@
+    file.read((char *)&id, 8);
+    file.read((char *)&isevidence, 1);
+    file.read((char *)&initial_value, 8);
+    file.read((char *)&type, 2);
+    file.read((char *)&edge_count, 8);
+    if (!file.read((char *)&cardinality, 8)) break;
+
+    // convert endian
+    id = be64toh(id);
+    type = be16toh(type);
+    long long tmp = be64toh(*(uint64_t *)&initial_value);
+    initial_value = *(double *)&tmp;
+    edge_count = be64toh(edge_count);
+    cardinality = be64toh(cardinality);
+
+    dprintf(
+        "----- id=%lli isevidence=%d initial=%f type=%d edge_count=%lli"
+        "cardinality=%lli\n",
+        id, isevidence, initial_value, type, edge_count, cardinality);
+
+    ++count;
+
+    variable_domain_type_t type_const;
+    switch (type) {
+      case 0:
+        type_const = DTYPE_BOOLEAN;
+        break;
+      case 1:
+        type_const = DTYPE_MULTINOMIAL;
+        break;
+      default:
+        std::cerr
+            << "[ERROR] Only Boolean and Multinomial variables are supported "
+               "now!"
+            << std::endl;
+        std::abort();
+    }
+    bool is_evidence = isevidence >= 1;
+    bool is_observation = isevidence == 2;
+    double init_value = is_evidence ? initial_value : 0;
+
+    variables[id] =
+        RawVariable(id, type_const, is_evidence, cardinality, init_value,
+                    init_value, edge_count, is_observation);
+
+    ++size.num_variables;
+    if (is_evidence) {
+      ++size.num_variables_evidence;
+    } else {
+      ++size.num_variables_query;
+    }
+  }
+  file.close();
+}
+
+void FactorGraph::load_factors(const std::string &filename) {
+  std::ifstream file;
+  file.open(filename.c_str(), std::ios::in | std::ios::binary);
+  size_t count = 0;
+  variable_id_t variable_id;
+  weight_id_t weightid;
+  variable_value_t value_id;
+  short type;
+  size_t edge_count;
+  variable_value_t equal_predicate;
+  bool ispositive;
+  while (file.good()) {
+    file.read((char *)&type, 2);
+    file.read((char *)&equal_predicate, 8);
+    if (!file.read((char *)&edge_count, 8)) break;
+
+    type = be16toh(type);
+    edge_count = be64toh(edge_count);
+    equal_predicate = be64toh(equal_predicate);
+
+    ++count;
+
+    factors[size.num_factors] = RawFactor(
+        size.num_factors, -1, (factor_function_type_t)type, edge_count);
+
+    for (size_t position = 0; position < edge_count; ++position) {
+      file.read((char *)&variable_id, 8);
+      file.read((char *)&ispositive, 1);
+      variable_id = be64toh(variable_id);
+
+      // check for wrong id
+      assert(variable_id < capacity.num_variables && variable_id >= 0);
+
+      // add variables to factors
+      factors[size.num_factors].add_variable_in_factor(
+          VariableInFactor(variable_id, position, ispositive, equal_predicate));
+      variables[variable_id].add_factor_id(size.num_factors);
+    }
+    size.num_edges += edge_count;
+
+    switch (type) {
+      case (FUNC_SPARSE_MULTINOMIAL): {
```

Comment:
1. Maybe get rid of any remaining references to `FUNC_MULTINOMIAL` or `FUNC_DENSE_MULTINOMIAL`?
2. The more precise name is probably `FUNC_CATEGORICAL_CONJUNCTION` or the like. (Once we support mixing categorical with boolean in the head, we can retire this naming as well because it'll collapse into FUNC_AND or something).